### PR TITLE
move to new protocol structure

### PR DESCRIPTION
**Pull request by naterush**
Based on spec in this issue: https://github.com/ethereum/cbc-casper/issues/175. Not finished yet, but opening for initial feedback.

For now, removing code for experiments and analysis. Will reintroduce this in future PR to be compatible with new protocol classes.

TODO:
- Removed some tests so I could get easier fix others. Will add them back in before final merge.

### CODE REVIEW

File: simulations/network_delay.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+def random_delay(sender, reciever, round):
+    """Messages take a random amount of time to deliver"""
+    return random.randint(0, CONSTANT)
+
+
+def gaussian_delay(sender, reciever, round):
+    """Messages take a random amount of time to deliver
+    (from a gaussian distribution)"""
+    return round(random.gauss(MU, SIGMA))
```
Comment (KentShikama) [inline simulations/network_delay.py]: See https://github.com/ethereum/cbc-casper/pull/180

File: casper/protocols/concurrent/concurrent_view.py [hunk 2]
```diff
@@ -1,46 +1,39 @@
...
+        self.create_outputs = create
+
     def estimate(self):
         """Returns the current forkchoice in this view"""
         available_outputs, output_sources = forkchoice.get_fork_choice(
-            self.last_finalized_estimate,
             self.children,
             self.latest_messages
         )
+        print(available_outputs)
```
Comment (djrtwo) [inline casper/protocols/concurrent/concurrent_view.py]: remove print

File: casper/protocols/integer/integer_plot_tool.py [hunk 3]
```diff
@@ -71,15 +71,11 @@ def _update_message_labels(self, new_messages):
...
-
-            if latest_message in self.bet_fault_tolerance:
-                continue
-
-            oracle = CliqueOracle(latest_message, validator.view, self.validator_set)
-            fault_tolerance, num_node_ft = oracle.check_estimate_safety()
-
-            if fault_tolerance > 0:
-                self.bet_fault_tolerance[latest_message] = num_node_ft
+        if self.view:
```
Comment (djrtwo) [inline casper/protocols/integer/integer_plot_tool.py]: When does not have a `view`?

File: casper/protocols/concurrent/concurrent_view.py (hunk 2 above)
Comment (djrtwo) [reply casper/protocols/concurrent/concurrent_view.py]: or is this part of the plot? If it is, move to plottool

File: simulations/exe_str_generator.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    def all_creators():
+        return range(num_validators)
+
+    return generate_execution(num_validators, num_rounds, network_delay_func, all_creators), num_validators
+
+
+def generate_rrob_execution(num_validators, num_rounds, network_delay_func):
+    curr_creator = 0
+    def rrob_creator():
+        nonlocal curr_creator
```
Comment (djrtwo) [inline simulations/exe_str_generator.py]: Nice, never seen `nonlocal` before

File: simulations/exe_str_generator.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
+"""The simulution utils module ... """
```
Comment (djrtwo) [inline simulations/exe_str_generator.py]: fix comment

File: tests/simulations/test_network.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+from simulations.network_delay import (
+    no_delay,
+    step_delay,
+    constant_delay,
+    random_delay,
+    gaussian_delay,
+    CONSTANT
+)
+
+def test_meets_interface(delay_function):
```
Comment (djrtwo) [inline tests/simulations/test_network.py]: two new lines between function defs

File: casper/protocol.py [hunk 7]
```diff
@@ -1,9 +1,136 @@
...
+    def _messages_needed_to_justify(self, message, validator):
+        """Returns the set of messages needed to justify a message to a validator"""
+        messages_needed = set([message])
+
+        current_message_hashes = set()
+        for message_hash in message.justification.values():
+            if message_hash not in validator.view.justified_messages:
+                current_message_hashes.add(message_hash)
+
+        while any(current_message_hashes):
```
[the remark this answers is not in the document]
Comment (naterush) [reply casper/protocol.py]: I think so! The duplicated code bothered me as well...

File: casper.py [hunk 8]
```diff
@@ -1,105 +1,97 @@
...
-        message_mode,
-        protocol=protocol,
-        network=network,
-        total_rounds=args.rounds,
-        report_interval=args.report_interval,
-        display=(not args.hide_display),
-        save=args.save,
+    # not all parameters are needed to generate the json for execution
+    execution_string = generate_json(**{k: v for k, v in vars(args).items() if k not in NOT_NEEDED})
+    print(type(execution_string))
```
[the remark this answers is not in the document]
Comment (naterush) [reply casper.py]: I changed it to print the execution string - which I think we want (so you can duplicate the execution if you want). 

Not sure why I was printing the type :)

File: casper/protocols/integer/integer_plot_tool.py (hunk 3 above)
Comment (naterush) [reply casper/protocols/integer/integer_plot_tool.py]: Good catch. This was an issue in a previous version of the architecture, but I fixed it. Removing!

**Comment by naterush** [comment]
@djrtwo Thanks for the review! Addressed your comments. I'm going to merge this and then do a release (going to skip the sharding release).

**Pull request merged**